Sqlalchemy Task

max-hoffman opened the pull request:
# TL;DR

This PR adds an `SQLAlchemyTask` modeled after `SQLite3Task`. This task loads a DataFrame from a database using an [connection dialect URL](https://docs.sqlalchemy.org/en/14/core/engines.html#database-urls).

Note:
- This is not Dolt (or any other database) specific, I tested with Dolt because that was easiest for me. This should be tested with MySQL, Postgres and SQLite at some point.

Optional:
- SQLAlchemy connection arguments can differ between dialects. A generic `connect_args` is provided to override the URL-specified values.
- I've added a naive secret loading option -- a dictionary of key->(name, group) pairs that will be merged with the connection args. Ex: `{"password": {"group": SECRET_GROUP, "name": SECRET_NAME}}` will override the password value with the result fetched from the Security Context.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
This is the entirety of the connection point. SQLAlchemy has lots of gotchas for different dialects, but those have to be shaken out in testing: 
```python
def execute(self, **kwargs) -> typing.Any:
    engine = create_engine(self._uri, connect_args=self._connect_args, echo=False)
    with engine.begin() as connection:
    df = pd.read_sql_query(self.get_query(**kwargs), connection)
```

I've tested this with MySQL. Choosing a multi-database testing setup is messy and different between librariese, and I have not added this to the PR yet. I usually see libraries matrix testing code like this, for example with docker compose setups that swap image based on database name, or using SQLAlchemy's ORM model to insert test data arbitrary to the database endpoint.

If I was going to add these, I would try to use [pytest database fixtures](https://pythonhosted.org/pytest-dbfixtures/howtouse.html) to get as close to unit-tests, before adding smoke tests for Task-runs in the orchestrator.

This is the only other non-boilerplate:
```python
        for key, secret in task_config.secret_connect_args.items():
            if "name" in secret and "group" in secret:
                value = flytekit.current_context().secrets.get(secret["group"], secret["name"])
                self._connect_args[key] = value
```

## Tracking Issue
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

kumare3 commented on flytekit/extras/sqlalchemy/task.py:
```diff
@@ -0,0 +1,73 @@
+        super().__init__(
+            name=name,
+            task_config=task_config,
+            task_type=self._SQLITE_TASK_TYPE,
```
task type needs to be updated

kumare3 commented on flytekit/extras/sqlalchemy/task.py:
```diff
@@ -0,0 +1,73 @@
+        uri: default sqlalchemy connector
+    """
+
+    uri: str
```
where do we get the username + password.  I think we should use 
Flyte secrets to access it (can be optional ofcourse)
https://docs.flyte.org/projects/cookbook/en/latest/auto_core_intermediate/use_secrets.html#use-secrets-in-a-task

kumare3 commented on flytekit/extras/sqlalchemy/task.py:
```diff
@@ -0,0 +1,84 @@
+        if task_config.secret_connect_args is not None:
+            for key, secret in task_config.secret_connect_args.items():
+                if "name" in secret and "group" in secret:
+                    value = current_context().secrets.get(secret["group"], secret["name"])
```
you can only access the secrets in the `execute` command or logically. They will be available here, but you should be using them only in the execute. Reason is __init__ will be invoked at convert / compile time, but execute only at runtime.

Also you will need to pass the secret requests to thparent 
https://github.com/flyteorg/flytekit/blob/master/flytekit/core/python_auto_container.py#L47

super().__init__(..., secret_requests=[...])

and access them in execute

max-hoffman replied:
got it. the secret_requests should pass through in kwargs

kumare3 commented on tests/flytekit/unit/extras/sqlalchemy/test_task.py:
```diff
@@ -0,0 +1,98 @@
+from flytekit import current_context, kwtypes, Secret, task, workflow
+from flytekit.extras.sqlalchemy.task import SQLAlchemyConfig, SQLAlchemyTask
+
+# https://www.sqlitetutorial.net/sqlite-sample-database/
```
delete?

max-hoffman replied:
yep

kumare3 commented on tests/flytekit/unit/extras/sqlalchemy/test_task.py:
```diff
@@ -0,0 +1,98 @@
+    try:
+        d = tempfile.TemporaryDirectory()
+        db_path = os.path.join(d.name, "tracks")
+        db = dolt.Dolt.init(db_path)
```
interesting that dolt is only a test dependencyy

max-hoffman replied:
this integration will work with any sqlalchemy dialect. Dolt is probably not the best way to test this, honestly. Setting up docker-compose resources for mysql, postgres, and a few others to test the integration is a lot of code, though https://haseebmajid.dev/blog/gitlab-ci-pytest-and-docker-compose. Other alternatives work, that's how i'd do it though.

kumare3 commented:
@max-hoffman so this is a Linux Foundation guideline, you have to signoff every commit you have using commit `-s`. One option could be to delete this branch and recreate a new branch that is a squash of all the commit with the `signoff` or use force signing every commit as explained in the DCO tab

max-hoffman commented:
@kumare3 I rebased with a signoff according to the DCO guidelines last night, is there additional steps?

kumare3 commented on flytekit/extras/sqlalchemy/task.py:
```diff
@@ -0,0 +1,84 @@
+
+    uri: str
+    connect_args: typing.Optional[typing.Dict[str, typing.Any]] = None
+    secret_connect_args: typing.Optional[typing.Dict[str, typing.Dict[str, typing.Any]]] = None
```
why not another dataclass? Its always 2 values at max right? Actual you can just use the secrets class here - https://github.com/flyteorg/flytekit/blob/8528268a29a07fe7e9ce9f7f08fea68c41b6a60b/flytekit/models/security.py#L11 right?

kumare3 commented on flytekit/extras/sqlalchemy/task.py:
```diff
@@ -0,0 +1,84 @@
+        outputs = kwtypes(results=output_schema_type if output_schema_type else FlyteSchema)
+        self._uri = task_config.uri
+        self._connect_args = task_config.connect_args or {}
+        self._secret_connect_args = task_config.secret_connect_args
```
I think we should patch the secrets from here into the parent constructor and then we can easily drop **kwargs.
Let me know if you want to chat about this, and maybe the model we have needs a little bit of an update to clean it up

max-hoffman replied:
if we're using `secret_requests` for config overrides we'll still need the key mapping somehow, unless we assume one of the Secret fields is the key

kumare3 commented:
@max-hoffman Sorry I did not realize this but for all new plugins we actually use the Microlib model. Can you please move the SQLAlchemy plugin to - https://github.com/flyteorg/flytekit/tree/master/plugins
This way all the dependencies and requirements are separately maintained. Also flytekitplugins-sql will be a separate package in pypi

max-hoffman commented:
@kumare3 Yeah no worries, I was wondering why the SQLite3 task was separated from other plugins.

kumare3 commented:
> @kumare3 Yeah no worries, I was wondering why the SQLite3 task was separated from other plugins.

ya just because it does not need extra libraries. We wanted a canonical SQL example to ship with Flytekit

wild-endeavor commented on .github/workflows/pythonbuild.yml:
```diff
@@ -35,7 +35,15 @@ jobs:
         make -C plugins install-all-dev
+        sudo bash -c 'curl -L https://github.com/dolthub/dolt/releases/latest/download/install.sh | sudo bash'
         pip freeze
+    - name: Install Dolt
```
Can we remove this?  Can we make this part of the plugin test?

max-hoffman replied:
This is the "which database to test with" trade-off. Testing with Dolt was easy but I need this binary. I think this should be replaced by more comprehensive tests with Postgres and Mysql, but those would have a bigger testing footprint and a longer dev-timeline.

max-hoffman replied:
maybe the plugin could be tested with SQLite, which wouldn't have dependency issues, but that might take me a bit to figure out

wild-endeavor replied:
You're right that we should run plugin unit tests as part of CI but that gets kicked off from the [Makefile](https://github.com/flyteorg/flytekit/blob/master/Makefile#L39) which depends on this [setup](https://github.com/flyteorg/flytekit/blob/master/.github/workflows/pythonbuild.yml#L37) line which then runs this pip [install](https://github.com/flyteorg/flytekit/blob/master/plugins/Makefile#L7) so the sqlalchemy setup.py is still run.  Any setup steps you put there will be executed.  It sounds like the implication is that pip installing this plugin isn't quite enough to get started, there's a few more commands to run after the pip install step before the user can start using it?  The [constructs around post-install](https://stackoverflow.com/questions/20288711/post-install-script-with-python-setuptools) hooks seem scarce though.  Can we replace with default behavior by any chance?

I don't know much about sqlalchemy...  is this plugin for all sqlalchemy?  Or is it only for dolt?  If there are hard dependencies on dolt for this plugin, should we rename to DoltSqlAlchemy?

Sorry for all the back and forth!  I'm out most of tomorrow (tuesday) but can do a google meet in the morning or the day after if maybe that's faster?  And thank you for doing this!

max-hoffman replied:
no worries! [this doc](https://docs.google.com/document/d/147nNlWrEJz7zLfIrr4rkokUceP8wFiCJuxIm2sS8ZZs/edit?usp=sharing) is my summary for what sqlalchemy is as a short summary. feel free to ping me in slack or discord to meet whenever. will follow up with specific QA below.

max-hoffman replied:
sqlalchemy is generic, not just for dolt. you put any SQL server string in there, specify the dialect (mysql, postgres, etc) and it will issue the proper select syntax. if you look at the code in  PR summary above, it's quite concise -- a connect string and a pandas read.

max-hoffman replied:
Setting up test database servers for every example data source is much more difficult. Each database server requires a different package except sqlite, maybe. That's why I added dolt as a test dependency -- to make a database server test fixture. I could have used postgres, mysql, oracle or sqlite instead.

max-hoffman replied:
I'm quickest with dolt -- in retrospect sqlite might have been a better choice. Ketan said you guys would have a way to inject server dependencies soon, so I assumed my tests would be phased out.

max-hoffman replied:
Post-install hooks would include the test dependencies in the package, which I think you want to avoid.

max-hoffman replied:
Thanks for taking so much time looking at this! I got the initial version working in 20 minutes but I really should have used sqlite -- saved us both from the back and forth

wild-endeavor commented on dev-requirements.in:
```diff
@@ -2,6 +2,7 @@
 
 black
 coverage[toml]
+doltcli
```
Can we remove this?

max-hoffman replied:
would a test-dependency in the plugin pull it in?

wild-endeavor replied:
I guess I'm a little confused... you're saying that users of the sqlalchemy plugin will need to install dolt dependencies as well to use it?  If that's the case, can we just move this line into the sqlalchemy setup.py?

max-hoffman replied:
sqlalchemy is database-independent, but needs at least one for testing. Moving to the sqlalchemy setup is OK but it's a test dependency -- if it's in the sqlalchemy setup.py test deps will it actually be installed for tests? (I think no? but I don't completely understand the micro-package organization.) I probably should have done this all with sqlite, it will take me awhile to rewrite the tests though

wild-endeavor commented on plugins/sqlalchemy/setup.py:
```diff
@@ -0,0 +1,35 @@
+
+microlib_name = f"flytekitplugins-{PLUGIN_NAME}"
+
+plugin_requires = ["flytekit>=0.16.0b0,<1.0.0", "sqlalchemy>=1.4.7", "pymysql>=1.0.2"]
```
```suggestion
plugin_requires = ["flytekit>=0.17.0,<1.0.0", "sqlalchemy>=1.4.7", "pymysql>=1.0.2"]
```

wild-endeavor commented on setup.py:
```diff
@@ -19,8 +19,9 @@
 hive_sensor = ["hmsclient>=0.0.1,<1.0.0"]
 notebook = ["papermill>=1.2.0", "nbconvert>=6.0.7", "ipykernel>=5.0.0"]
 sagemaker = ["sagemaker-training>=3.6.2,<4.0.0"]
+mysql = ["sqlalchemy>=1.4.7", "pymysql>=1.0.2"]
```
let's nix this line?  These are all for the old style of plugins (the ones that live in flytekit/plugins rather than the new top level plugins folder).  We're trying to deprecate all the old-api related things.

wild-endeavor commented on setup.py:
```diff
@@ -19,8 +19,9 @@
+mysql = ["sqlalchemy>=1.4.7", "pymysql>=1.0.2"]
 
-all_but_spark = sidecar + schema + hive_sensor + notebook + sagemaker
+all_but_spark = sidecar + schema + hive_sensor + notebook + sagemaker + mysql
```
revert here as well.

wild-endeavor commented on plugins/tests/sqlalchemy/test_task.py:
```diff
@@ -0,0 +1,96 @@
+    try:
+        d = tempfile.TemporaryDirectory()
+        db_path = os.path.join(d.name, "tracks")
+        db = dolt.Dolt.init(db_path)
```
Could we just remove this test, and then remove all dolt-specific dependencies?  Then we should merge.  I'm totally happy having a dolt-specific example in the flytesnacks repository under [cookbook/plugins](https://github.com/flyteorg/flytesnacks/tree/master/cookbook/plugins) that uses this task.  I think that would be cleanest.

max-hoffman commented:
@wild-endeavor I removed dolt tests, added an sqlite test because that doesn't require dependencies, but coverage is going to be missing lines because the secret-parsing paths aren't checked. I can add code that hits the secret lines with sqlite, but without a way to check whether they were set correctly.

wild-endeavor commented on plugins/sqlalchemy/setup.py:
```diff
@@ -0,0 +1,35 @@
+        "Topic :: Software Development :: Libraries",
+        "Topic :: Software Development :: Libraries :: Python Modules",
+    ],
+    scripts=["scripts/flytekit_install_sqlalchemy.sh"],
```
okay one last comment sorry.  this script is missing.  do we need this?  Can we move the dolt install script to the example in the future as well and just remove it, or is there a new script that's needed?

max-hoffman replied:
no script needed, good catch i'll fix in a couple

wild-endeavor commented on plugins/sqlalchemy/scripts/flytekit_install_dolt.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
should we get rid of this as well?

max-hoffman replied:
yeah, sorry this is just a failure of detail-attentiveness on my part, the size of this PR discussion is getting out of hand. i'll make everything cleaner next time if you guys end up liking the dolt plugin PR

wild-endeavor replied:
oh no not at all, thank you for putting up with me!  Thanks for figuring out the sqlite test too.
